File: examples/Algorithm/python/05_sorting.py
"""
Sorting Algorithms
Sorting Algorithms Comparison

Implementation and comparison of various sorting algorithms.
"""

import random
import time
from typing import List


# =============================================================================
# 1. Bubble Sort
# =============================================================================
def bubble_sort(arr: List[int]) -> List[int]:
    """
    Bubble Sort
    Time: O(n^2), Space: O(1), Stable
    """
    arr = arr.copy()
    n = len(arr)

    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped:
            break

    return arr


# =============================================================================
# 2. Selection Sort
# =============================================================================
def selection_sort(arr: List[int]) -> List[int]:
    """
    Selection Sort
    Time: O(n^2), Space: O(1), Unstable
    """
    arr = arr.copy()
    n = len(arr)

    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

    return arr


# =============================================================================
# 3. Insertion Sort
# =============================================================================
def insertion_sort(arr: List[int]) -> List[int]:
    """
    Insertion Sort
    Time: O(n^2), Best O(n), Space: O(1), Stable
    Efficient for nearly sorted data
    """
    arr = arr.copy()
    n = len(arr)

    for i in range(1, n):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

    return arr


# =============================================================================
# 4. Merge Sort
# =============================================================================
def merge_sort(arr: List[int]) -> List[int]:
    """
    Merge Sort
    Time: O(n log n), Space: O(n), Stable
    """
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


def merge(left: List[int], right: List[int]) -> List[int]:
    """Merge two sorted arrays"""
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result


# =============================================================================
# 5. Quick Sort
# =============================================================================
def quick_sort(arr: List[int]) -> List[int]:
    """
    Quick Sort
    Time: Average O(n log n), Worst O(n^2), Space: O(log n), Unstable
    """
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quick_sort(left) + middle + quick_sort(right)


def quick_sort_inplace(arr: List[int], low: int = 0, high: int = None) -> List[int]:
    """Quick Sort (in-place)"""
    if high is None:
        arr = arr.copy()
        high = len(arr) - 1

    if low < high:
        pivot_idx = partition(arr, low, high)
        quick_sort_inplace(arr, low, pivot_idx - 1)
        quick_sort_inplace(arr, pivot_idx + 1, high)

    return arr


def partition(arr: List[int], low: int, high: int) -> int:
    """Partition (Lomuto scheme)"""
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


# =============================================================================
# 6. Heap Sort
# =============================================================================
def heap_sort(arr: List[int]) -> List[int]:
    """
    Heap Sort
    Time: O(n log n), Space: O(1), Unstable
    """
    arr = arr.copy()
    n = len(arr)

    # Build max heap
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Extract one by one
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)

    return arr


def heapify(arr: List[int], n: int, i: int):
    """Maintain max heap property"""
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


# =============================================================================
# 7. Counting Sort
# =============================================================================
def counting_sort(arr: List[int]) -> List[int]:
    """
    Counting Sort
    Time: O(n + k), Space: O(k), Stable
    k = max_value - min_value + 1
    Efficient when the integer range is small
    """
    if not arr:
        return []

    min_val, max_val = min(arr), max(arr)
    range_val = max_val - min_val + 1

    count = [0] * range_val
    output = [0] * len(arr)

    # Count
    for num in arr:
        count[num - min_val] += 1

    # Accumulate
    for i in range(1, range_val):
        count[i] += count[i - 1]

    # Place in reverse order (to maintain stability)
    for i in range(len(arr) - 1, -1, -1):
        num = arr[i]
        count[num - min_val] -= 1
        output[count[num - min_val]] = num

    return output


# =============================================================================
# 8. Radix Sort
# =============================================================================
def radix_sort(arr: List[int]) -> List[int]:
    """
    Radix Sort (LSD)
    Time: O(d * (n + k)), Space: O(n + k)
    d = number of digits, k = radix (usually 10)
    Version without negative number support
    """
    if not arr or min(arr) < 0:
        return sorted(arr)  # Fall back to default sort for negative numbers

    max_val = max(arr)

    exp = 1
    while max_val // exp > 0:
        arr = counting_sort_by_digit(arr, exp)
        exp *= 10

    return arr


def counting_sort_by_digit(arr: List[int], exp: int) -> List[int]:
    """Counting sort by a specific digit"""
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for num in arr:
        digit = (num // exp) % 10
        count[digit] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        digit = (arr[i] // exp) % 10
        count[digit] -= 1
        output[count[digit]] = arr[i]

    return output


# =============================================================================
# Performance Comparison
# =============================================================================
def benchmark_sorts(size: int = 1000):
    """Sorting algorithm performance comparison"""
    arr = [random.randint(0, 10000) for _ in range(size)]

    algorithms = [
        ("Bubble Sort", bubble_sort, size <= 1000),
        ("Selection Sort", selection_sort, size <= 1000),
        ("Insertion Sort", insertion_sort, size <= 1000),
        ("Merge Sort", merge_sort, True),
        ("Quick Sort", quick_sort, True),
        ("Heap Sort", heap_sort, True),
        ("Counting Sort", counting_sort, True),
        ("Radix Sort", radix_sort, True),
        ("Python sorted()", sorted, True),
    ]

    print(f"\nArray Size: {size}")
    print("-" * 50)

    for name, func, should_run in algorithms:
        if should_run:
            test_arr = arr.copy()
            start = time.perf_counter()
            result = func(test_arr)
            elapsed = time.perf_counter() - start
            print(f"{name:20s}: {elapsed * 1000:8.3f} ms")
        else:
            print(f"{name:20s}: (skipped due to size)")


# =============================================================================
# Tests
# =============================================================================
def main():
    print("=" * 60)
    print("Sorting Algorithms")
    print("=" * 60)

    # Test array
    arr = [64, 34, 25, 12, 22, 11, 90]
    print(f"\nOriginal array: {arr}")
    print("-" * 40)

    # Test each algorithm
    algorithms = [
        ("Bubble Sort", bubble_sort),
        ("Selection Sort", selection_sort),
        ("Insertion Sort", insertion_sort),
        ("Merge Sort", merge_sort),
        ("Quick Sort", quick_sort),
        ("Heap Sort", heap_sort),
        ("Counting Sort", counting_sort),
        ("Radix Sort", radix_sort),
    ]

    for name, func in algorithms:
        result = func(arr.copy())
        print(f"{name}: {result}")

    # Performance Comparison
    print("\n" + "=" * 60)
    print("Performance Benchmark")
    print("=" * 60)

    for size in [100, 1000, 5000]:
        benchmark_sorts(size)

    # Sorting Algorithm Comparison Table
    print("\n" + "=" * 60)
    print("Sorting Algorithm Comparison")
    print("=" * 60)
    print("""
    | Algorithm       | Average  | Worst    | Space  | Stable | Notes                        |
    |----------------|----------|----------|--------|--------|------------------------------|
    | Bubble Sort    | O(n^2)   | O(n^2)   | O(1)   | Yes    | Simple, educational          |
    | Selection Sort | O(n^2)   | O(n^2)   | O(1)   | No     | Simple, fewer swaps          |
    | Insertion Sort | O(n^2)   | O(n^2)   | O(1)   | Yes    | Good for nearly sorted data  |
    | Merge Sort     | O(nlogn) | O(nlogn) | O(n)   | Yes    | Consistent performance       |
    | Quick Sort     | O(nlogn) | O(n^2)   | O(logn)| No     | Fastest on average           |
    | Heap Sort      | O(nlogn) | O(nlogn) | O(1)   | No     | In-place, consistent perf    |
    | Counting Sort  | O(n+k)   | O(n+k)   | O(k)   | Yes    | Integers, small range        |
    | Radix Sort     | O(d(n+k))| O(d(n+k))| O(n+k) | Yes    | Digit-based                  |

    Practical Selection Guide:
    - General case: Quick Sort or built-in language sort
    - Stability needed: Merge Sort
    - Memory constrained: Heap Sort
    - Nearly sorted: Insertion Sort
    - Integers + small range: Counting/Radix Sort
    """)


if __name__ == "__main__":
    main()

File: examples/Algorithm/cpp/05_sorting.cpp
/*
 * Sorting Algorithms
 * Bubble, Selection, Insertion, Merge, Quick, Heap, Counting, Radix
 *
 * Implementation and comparison of various sorting algorithms.
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <cmath>

using namespace std;

// =============================================================================
// 1. Bubble Sort - O(n^2)
// =============================================================================

void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }
        if (!swapped) break;  // Optimization
    }
}

// =============================================================================
// 2. Selection Sort - O(n^2)
// =============================================================================

void selectionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        swap(arr[i], arr[minIdx]);
    }
}

// =============================================================================
// 3. Insertion Sort - O(n^2)
// =============================================================================

void insertionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

// =============================================================================
// 4. Merge Sort - O(n log n)
// =============================================================================

void merge(vector<int>& arr, int left, int mid, int right) {
    vector<int> L(arr.begin() + left, arr.begin() + mid + 1);
    vector<int> R(arr.begin() + mid + 1, arr.begin() + right + 1);

    size_t i = 0, j = 0;
    int k = left;

    while (i < L.size() && j < R.size()) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while (i < L.size()) arr[k++] = L[i++];
    while (j < R.size()) arr[k++] = R[j++];
}

void mergeSort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

// =============================================================================
// 5. Quick Sort - O(n log n) average
// =============================================================================

int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// 3-way Quick Sort (efficient for duplicate elements)
void quickSort3Way(vector<int>& arr, int low, int high) {
    if (low >= high) return;

    int lt = low, gt = high;
    int pivot = arr[low];
    int i = low + 1;

    while (i <= gt) {
        if (arr[i] < pivot) {
            swap(arr[lt++], arr[i++]);
        } else if (arr[i] > pivot) {
            swap(arr[i], arr[gt--]);
        } else {
            i++;
        }
    }

    quickSort3Way(arr, low, lt - 1);
    quickSort3Way(arr, gt + 1, high);
}

// =============================================================================
// 6. Heap Sort - O(n log n)
// =============================================================================

void heapify(vector<int>& arr, int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

void heapSort(vector<int>& arr) {
    int n = arr.size();

    // Build heap
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    // Extract one by one
    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

// =============================================================================
// 7. Counting Sort - O(n + k)
// =============================================================================

void countingSort(vector<int>& arr) {
    if (arr.empty()) return;

    int maxVal = *max_element(arr.begin(), arr.end());
    int minVal = *min_element(arr.begin(), arr.end());
    int range = maxVal - minVal + 1;

    vector<int> count(range, 0);
    vector<int> output(arr.size());

    for (int x : arr) {
        count[x - minVal]++;
    }

    for (int i = 1; i < range; i++) {
        count[i] += count[i - 1];
    }

    for (int i = arr.size() - 1; i >= 0; i--) {
        output[count[arr[i] - minVal] - 1] = arr[i];
        count[arr[i] - minVal]--;
    }

    arr = output;
}

// =============================================================================
// 8. Radix Sort - O(d * (n + k))
// =============================================================================

void countingSortForRadix(vector<int>& arr, int exp) {
    int n = arr.size();
    vector<int> output(n);
    vector<int> count(10, 0);

    for (int x : arr) {
        count[(x / exp) % 10]++;
    }

    for (int i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    for (int i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
    }

    arr = output;
}

void radixSort(vector<int>& arr) {
    if (arr.empty()) return;

    int maxVal = *max_element(arr.begin(), arr.end());

    for (int exp = 1; maxVal / exp > 0; exp *= 10) {
        countingSortForRadix(arr, exp);
    }
}

// =============================================================================
// Test
// =============================================================================

void printVector(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); i++) {
        cout << v[i];
        if (i < v.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    cout << "============================================================" << endl;
    cout << "Sorting Algorithm Examples" << endl;
    cout << "============================================================" << endl;

    vector<int> original = {64, 34, 25, 12, 22, 11, 90};

    // 1. Bubble Sort
    cout << "\n[1] Bubble Sort" << endl;
    vector<int> arr1 = original;
    bubbleSort(arr1);
    cout << "    Result: ";
    printVector(arr1);
    cout << endl;

    // 2. Selection Sort
    cout << "\n[2] Selection Sort" << endl;
    vector<int> arr2 = original;
    selectionSort(arr2);
    cout << "    Result: ";
    printVector(arr2);
    cout << endl;

    // 3. Insertion Sort
    cout << "\n[3] Insertion Sort" << endl;
    vector<int> arr3 = original;
    insertionSort(arr3);
    cout << "    Result: ";
    printVector(arr3);
    cout << endl;

    // 4. Merge Sort
    cout << "\n[4] Merge Sort" << endl;
    vector<int> arr4 = original;
    mergeSort(arr4, 0, arr4.size() - 1);
    cout << "    Result: ";
    printVector(arr4);
    cout << endl;

    // 5. Quick Sort
    cout << "\n[5] Quick Sort" << endl;
    vector<int> arr5 = original;
    quickSort(arr5, 0, arr5.size() - 1);
    cout << "    Result: ";
    printVector(arr5);
    cout << endl;

    // 6. Heap Sort
    cout << "\n[6] Heap Sort" << endl;
    vector<int> arr6 = original;
    heapSort(arr6);
    cout << "    Result: ";
    printVector(arr6);
    cout << endl;

    // 7. Counting Sort
    cout << "\n[7] Counting Sort" << endl;
    vector<int> arr7 = original;
    countingSort(arr7);
    cout << "    Result: ";
    printVector(arr7);
    cout << endl;

    // 8. Radix Sort
    cout << "\n[8] Radix Sort" << endl;
    vector<int> arr8 = {170, 45, 75, 90, 802, 24, 2, 66};
    cout << "    Original: ";
    printVector(arr8);
    cout << endl;
    radixSort(arr8);
    cout << "    Result: ";
    printVector(arr8);
    cout << endl;

    // 9. Complexity Comparison
    cout << "\n[9] Sorting Algorithm Complexity Comparison" << endl;
    cout << "    | Algorithm  | Best      | Average   | Worst     | Space | Stable |" << endl;
    cout << "    |------------|-----------|-----------|-----------|-------|--------|" << endl;
    cout << "    | Bubble     | O(n)      | O(n^2)    | O(n^2)    | O(1)  | Yes    |" << endl;
    cout << "    | Selection  | O(n^2)    | O(n^2)    | O(n^2)    | O(1)  | No     |" << endl;
    cout << "    | Insertion  | O(n)      | O(n^2)    | O(n^2)    | O(1)  | Yes    |" << endl;
    cout << "    | Merge      | O(n log n)| O(n log n)| O(n log n)| O(n)  | Yes    |" << endl;
    cout << "    | Quick      | O(n log n)| O(n log n)| O(n^2)    | O(log)| No     |" << endl;
    cout << "    | Heap       | O(n log n)| O(n log n)| O(n log n)| O(1)  | No     |" << endl;
    cout << "    | Counting   | O(n+k)    | O(n+k)    | O(n+k)    | O(k)  | Yes    |" << endl;
    cout << "    | Radix      | O(d(n+k)) | O(d(n+k)) | O(d(n+k)) | O(n+k)| Yes    |" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}
